Sources/App/Controllers/PackageController+routes.swift

```diff
@@ -104,18 +105,59 @@ struct PackageController {
 
         let url = try Self.awsDocumentationURL(owner: owner, repository: repository, reference: reference, fragment: fragment, path: path)
         let awsResponse = try await Current.fetchDocumentation(req.client, url)
-        guard (200..<399).contains(awsResponse.status.code) else {
-            // Convert anything that isn't a 2xx or 3xx into a 404
-            return try await DocumentationErrorPage.View(path: req.url.path,
-                                                         error: Abort(awsResponse.status))
-            .document()
-            .encodeResponse(status: .notFound,
-                            headers: req.headers.replacingOrAdding(name: .cacheControl,
-                                                                   value: "no-cache"),
-                            for: req)
+
+        // Never let a request continue if the AWS request fails, except to a potential `/owner/repo/ref/documentation/` redirect.
+        guard (200..<399).contains(awsResponse.status.code) || fragment == .documentationRedirect else {
+            // Convert anything that isn't a 2xx or 3xx from AWS into a 404 from us.
+            throw Abort(.notFound)
         }
 
         switch fragment {
+            case .documentationRedirect:
+                let referenceToMatch: Reference = {
+                    if let semanticVersion = SemanticVersion(reference) {
+                        return .tag(semanticVersion, reference)
+                    } else {
+                        return .branch(reference)
+                    }
+                }()
+
+                guard let queryResult = try await Joined3<Version, Package, Repository>
+                    .query(on: req.db,
+                           join: \Version.$package.$id == \Package.$id, method: .inner,
+                           join: \Package.$id == \Repository.$package.$id, method: .inner)
+                    .filter(Repository.self, \.$owner, .custom("ilike"), owner)
+                    .filter(Repository.self, \.$name, .custom("ilike"), repository)
+                    .filter(\Version.$reference == referenceToMatch)
```

Comment:
Is there a way I'm missing to do a JSON comparison here, @finestructure? The SQL that this outputs:

```
SELECT "versions"."doc_archives" AS "versions_doc_archives" FROM "versions" INNER JOIN "packages" ON "versions"."package_id" = "packages"."id" INNER JOIN "repositories" ON "packages"."id" = "repositories"."package_id" WHERE "repositories"."owner" ilike $1 AND "repositories"."name" ilike $2 AND "versions"."reference" = $3 AND "versions"."doc_archives" IS NOT NULL LIMIT 1 [Optional("owner"), Optional("package"), main]
```

So the `reference` where clause is simply comparing against `'main'`, but it's a JSON field.

Reply:
I'm pretty sure that's just rendering as `main` in the dump because of `Reference`'s `CustomStringConvertible` conformance. You're really comparing a `JSONB` field to a structured type `Reference`, which should work as expected. I believe we're doing this elsewhere as well.